src/main/java/com/cloudbees/plugins/credentials/CredentialsProvider.java

```diff
@@ -1519,51 +1523,55 @@ public static Fingerprint getOrCreateFingerprintOf(@NonNull Credentials c) throw
-                                    iterator.remove();
+                        Fingerprint fingerprint = getOrCreateFingerprintOf(c);
+                        BulkChange change = new BulkChange(fingerprint);
+                        try {
+                            Collection<FingerprintFacet> facets = fingerprint.getFacets();
+                            // purge any old facets
+                            long start = timestamp;
+                            for (Iterator<FingerprintFacet> iterator = facets.iterator(); iterator.hasNext(); ) {
+                                FingerprintFacet f = iterator.next();
+                                // For all the node-tracking credentials, check to see if we can remove
+                                // older instances of these credential fingerprints, or from nodes which no longer exist
+                                if (f instanceof NodeCredentialsFingerprintFacet) {
+                                    // Remove older instances
+                                    if (StringUtils.equals(nodeName, ((NodeCredentialsFingerprintFacet) f).getNodeName())) {
+                                        start = Math.min(start, f.getTimestamp());
+                                        iterator.remove();
+                                    // Remove unneeded instances
+                                    } else if (!jenkinsNodeNames.contains(((NodeCredentialsFingerprintFacet) f).getNodeName())) {
+                                        iterator.remove();
+                                    }
                                 }
                             }
+                            // add in the new one if it is a valid current node
+                            if (jenkinsNodeNames.contains(node.getNodeName())) {
+                                facets.add(new NodeCredentialsFingerprintFacet(node, fingerprint, start, timestamp));
+                            }
+                        } finally {
+                            change.commit();
                         }
-                        // add in the new one if it is a valid current node
-                        if (jenkinsNodeNames.contains(node.getNodeName())) {
-                            facets.add(new NodeCredentialsFingerprintFacet(node, fingerprint, start, timestamp));
-                        }
-                    } finally {
-                        change.commit();
+                    } catch (IOException e) {
+                        LOGGER.log(Level.FINEST, "Could not track usage of " + c, e);
                     }
-                } catch (IOException e) {
-                    LOGGER.log(Level.FINEST, "Could not track usage of " + c, e);
                 }
             }
+        } else {
+            LOGGER.log(Level.FINEST, "TrackAll(Node) called but fingerprints disabled in the global configuration.");
```

Comment:
I suppose the intention is to distinguish between the node and run methods, but it feels a little confusing to have part of the signature. (Not sure it's really intended to show the signature, but that's kind of what it looks like.)

Reply:
The intent was to distinguish between the node, the item and the run versions. I will change the format to avoid confusion with signature.